bundles/org.openhab.core/src/main/java/org/openhab/core/i18n/I18nException.java

```diff
@@ -0,0 +1,164 @@
+    }
+
+    /**
+     * Setup the internationalization support
+     *
+     * In case the message starts with "@text/", one of both setupI18n() methods needs to be called before using the
+     * getMessage() and getLocalizedMessage() methods.
+     *
+     * The getLocalizedMessage() method will then consider the language provided by the locale parameter.
+     *
+     * @param bundle the bundle containing the i18n properties
+     * @param i18nProvider the translation provider
+     * @param locale the language to use for localizing the message
+     */
+    public void setupI18n(Bundle bundle, TranslationProvider i18nProvider, Locale locale) {
+        this.bundle = bundle;
+        this.i18nProvider = i18nProvider;
+        this.locale = locale;
+    }
+
+    @Override
+    public @Nullable String getMessage() {
+        Bundle localeBundle = this.bundle;
+        TranslationProvider localI18nProvider = this.i18nProvider;
+        if (msgKey.isBlank()) {
+            return super.getMessage();
+        } else if (localeBundle == null || localI18nProvider == null) {
+            logger.warn("Internationalization support is not setup for the exception");
+            return super.getMessage();
+        } else {
+            return localI18nProvider.getText(localeBundle, msgKey, null, Locale.ENGLISH, msgParams);
+        }
+    }
+
+    @Override
+    public @Nullable String getLocalizedMessage() {
+        Bundle localeBundle = this.bundle;
+        TranslationProvider localI18nProvider = this.i18nProvider;
+        Locale localLocale = this.locale;
+        if (msgKey.isBlank()) {
+            return super.getMessage();
+        } else if (localeBundle == null || localI18nProvider == null) {
+            logger.warn("Internationalization support is not setup for the exception");
+            return super.getMessage();
+        } else {
+            return localI18nProvider.getText(localeBundle, msgKey, null,
+                    localLocale != null ? localLocale : Locale.ENGLISH, msgParams);
+        }
+    }
+
+    /**
+     * Provides the raw message
+     *
+     * If the message does not start with "@text/", it returns the same as the getMessage() method.
+     * If the message starts with "@text/" and no optional arguments are set, it returns a string of this
+     * kind: @text/key
+     * If the message starts with "@text/" and optional arguments are set, it returns a string of this kind: @text/key [
+     * "param1", "param2" ]
+     *
+     * @return the raw message or null if the message is undefined
+     */
+    public @Nullable String getRawMessage() {
+        if (msgKey.isBlank()) {
+            return super.getMessage();
+        }
+        String result = "@text/" + msgKey;
+        Object @Nullable [] params = msgParams;
+        if (params != null && params.length > 0) {
+            result += Stream.of(params).map(param -> String.format("\"%s\"", param == null ? "" : param.toString()))
```

Comment:
Do we need the call of `toString()` on the `param`? Doesn't `String.format(...)` takes care of that?

```suggestion
            result += Stream.of(params)filter(param -> param != null && !param.isBlank()).map(param -> String.format("\"%s\"", param))
```

Reply:
This can lead to parameters mismatch if you remove some of them ?

Reply:
Please correct me if I am wrong: My current understanding of this method is to create a result like `@text/key [ "param_1", "param_2", ... , "param_n" ]`. If one of the parameters is `null` you currently will get `@text/key [ "param_1", "", "param_3", ... , "param_n" ]`. Is that your expected result? Or do you expect `@text/key [ "param_1", "param_3", ... , "param_n" ]`?

Reply:
I am expecting `@text/key [ "param_1", "", "param_3", ... , "param_n" ]` to be sure matching the parameter patten from the properties file.
If in the properties file, I have a value like `example with {0} and {1}`. If my exception is instantiated with parameters null and "param2", I am expecting to have a message `example  and param2`, not `example param2 and {1}`.

Reply:
That's clear now. I finally understood how parametrized translations are working. Question solved.